modules/appeals_api/lib/appeals_api/stats_report.rb

```diff
@@ -0,0 +1,147 @@
+
+        #{formatted_stats(AppealsApi::SupplementalClaim)}
+        ### Stalled records
+
+        #{formatted_stalled(AppealsApi::SupplementalClaim)}
+      REPORT
+    end
+
+    private
+
+    attr_accessor :date_from, :date_to
+
+    def status_update_records(statusable_type, status_from, status_to)
+      @records ||= {}
+      @records[statusable_type] ||= {}
+      @records[statusable_type][status_from] ||= {}
+      @records[statusable_type][status_from][status_to] ||=
+        begin
+          records = AppealsApi::StatusUpdate.where(
+            from: status_from,
+            to: status_to,
+            statusable_type: statusable_type,
+            status_update_time: date_from..date_to
+          ).order(:statusable_id)
+          previous_records = AppealsApi::StatusUpdate.where(
+            to: status_from,
+            statusable_id: records.pluck(:statusable_id),
+            statusable_type: statusable_type
+          ).order(:statusable_id)
+          records.zip(previous_records)
+        end
+    end
+
+    def stats(update_record_pairs)
+      return { mean: nil, median: nil } if update_record_pairs.empty?
+
+      sum, values = update_record_pairs.reduce([0, []]) do |(s, v), (current, previous)|
+        timespan = current.status_update_time - previous.status_update_time
+        [s + timespan, v << timespan]
+      end
+
+      values.sort!
+      middle = (values.count - 1) / 2.0
+
+      {
+        mean: sum / values.count,
+        median: (values[middle.floor] + values[middle.ceil]) / 2.0
+      }
+    end
+
+    def stalled_records(appeal_class, status)
+      @stalled ||= {}
+      @stalled[appeal_class.name] ||= {}
+      @stalled[appeal_class.name][status] ||= appeal_class.where(
+        'updated_at < ?',
+        (date_to - STALLED_THRESHOLD).beginning_of_day
+      ).where(status: status).order(created_at: :desc)
+    end
+
+    def timespan_in_words(seconds)
+      return '(none)' if seconds.nil?
+
+      minutes, = seconds.divmod(60)
+      hours, minutes = minutes.divmod(60)
+      days, hours = hours.divmod(24)
+
+      "#{days}d #{hours}h #{minutes}m"
+    end
+
+    def formatted_stats(appeal_class)
+      parts = STATUS_TRANSITION_PAIRS.map do |(status_from, status_to)|
+        values = stats(status_update_records(appeal_class.name, status_from, status_to))
+
+        <<~STATS
+          From '#{status_from}' to '#{status_to}':
+          * Average: #{timespan_in_words(values[:mean])}
+          * Median:  #{timespan_in_words(values[:median])}
+        STATS
+      end
+
+      parts.join("\n")
+    end
+
+    def formatted_stalled(appeal_class)
+      parts = STATUS_TRANSITION_PAIRS.collect(&:first).uniq.map do |status|
+        lines = ["Stalled in '#{status}':"]
+        stalled = stalled_records(appeal_class, status)
+
+        if stalled.empty?
+          lines << '(none)'
+        else
+          stalled.each { |record| lines << "* #{record.id} (since #{record.updated_at.iso8601})" }
```

Comment:
We don't actually need another _list_ of stalled records - we have a couple other reports for that already. In this report, we are more interested in the statistics of those stalled records such as "how many records are older than 6 months? how many between 3-6 mo's" etc.

Reply:
Ok, I've updated the stalled section to look like this (this output is from the tests):
```
(omitted upper section with timespans)

### Stalled records

Stalled in 'processing':
* 4-5 months: 1
* 5-6 months: 2
* > 6 months: 1
Stalled in 'submitted':
* 3-4 months: 1
* 4-5 months: 2
* 5-6 months: 1
Stalled in 'error':
* 3-4 months: 2
* 4-5 months: 1
* 5-6 months: 1
```

If there are no stalled records it looks like this:
```
Stalled in 'processing':
(none)
Stalled in 'submitted':
(none)
Stalled in 'error':
(none)
```

Happy to update these timespans if wanted, too!

Reply:
Timelines are fine, I think. Let skip printing anything for statuses that didn't stall, though.